src/repository.c

```diff
@@ -2290,6 +2438,13 @@ int git_repository_set_head(
 	if (error < 0 && error != GIT_ENOTFOUND)
 		goto cleanup;
 
+	if (ref && current->type == GIT_REF_SYMBOLIC && git__strcmp(current->target.symbolic, ref->name)
+		&& git_reference_is_branch(ref) && git_branch_is_checked_out(ref))
+	{
```

Comment:
Stylistic comment:  the `{` should be pulled up to the line above...  There are a few occurrences of this - I did not point each one out, but I can if it would be helpful.